### PR TITLE
Safely recover from missing srcref during srcref matching

### LINKED ISSUE #27: Reexports crash the covtracer
While using the `covtracer` package, I've encountered a bug that crashes the `test_trace_df.coverage` function if the passes coverage file was built for a package that has both exports and reexports. I'll try to give a detailed look at the issue based on the example of a `geepack` package (https://cran.r-project.org/web/packages/geepack/index.html). 

To begin with, let's starts with the actual place where the error occurs, it is in the `match_containing_srcrefs` helper (https://github.com/Genentech/covtracer/blob/main/R/srcref_df.R#L274) and it is thrown because of the unexpected NA passed to `Position` function. That NA comes from the `rdf` object defined here https://github.com/Genentech/covtracer/blob/main/R/srcref_df.R#L262. Here is how problematic part of the data frame looks like:

```
                  name                                                                                                        srcfile line1 byte1 line2
24 print.summary.geese /private/var/folders/bh/g3p3b4sn0lx9ml5kr20bbq3c0000gp/T/RtmpWMzIhZ/R.INSTALL112c04f3c5fbb/geepack/R/summary.R    83    24   117
33                 %>%                                                                                                           <NA>    NA    NA    NA
34                tidy                                                                                                           <NA>    NA    NA    NA
   byte2 col1 col2 parsed1 parsed2
24     1   24    1    3003    3037
33    NA   NA   NA      NA      NA
34    NA   NA   NA      NA      NA
```

It is understandable why do we have NA there, `%>%` and `tidy` are reexports so there is no source for them within the package. 

On the question how is that situation different from the assertive pacakge, for which we can successfully execute this function, the answer is this line https://github.com/Genentech/covtracer/blob/main/R/srcref_df.R#L271. assertive has only reexports so the `ldf` object has 0 rows and we never enter that loop, meaning, w never touch line 274 that is crashing the function.

Last but not least I want to clarify how the stack looks like there:

`test_trace_df.coverage` -> `join_on_containing_srcrefs` (https://github.com/Genentech/covtracer/blob/main/R/test_trace_df.R#L92-L96) -> `match_containing_srcrefs` (https://github.com/Genentech/covtracer/blob/main/R/srcref_df.R#L342)

I'm pretty sure the issues come from that reexports thingy, looking at the objects that are passed downstream in `test_trace_df.coverage`, I just don't feel experienced enough to solve this issue on my own.

### PR DESCRIPTION
**Pull request by dgkf**
Resolves #27 (attempt 2 :P)

Code-wise, this is a pretty straightforward fix. It took me far longer to write up a test that tests for this edge case to make sure this doesn't break in the future.

Nevertheless, please take a look, @maksymiuks and let me know if these changes resolve the issue you found.

### CODE REVIEW

File: R/srcref_df.R [hunk 1]
```diff
@@ -258,6 +258,15 @@ test_trace_mapping <- function(x) {
...
 match_containing_srcrefs <- function(l, r) {
+  # NOTE:
+  #   srcrefs are matched due to filenames, as full file paths are often
+  #   different. Trace srcrefs srcfiles will point to a temporary installation
+  #   when running covr, whereas package namespace srcrefs srcfiles will point
+  #   to the package install location.
+  #
+  # TODO:
+  #   more comprehensive path comparisons (perhaps pkgname/R/file.R)
+
```
Comment (dgkf) [inline R/srcref_df.R:269]: @maksymiuks - I identified this potential pitfall when digging into this. Now that srcrefs might originate from other packages, just comparing basenames might not be sufficient. I also saw cases where srcrefs originated from the `inst` directory, which might have duplicated basenames also. 

This is probably an exceedingly rare edge case, but something we may want to think about.

File: R/srcref_df.R [hunk 2]
```diff
@@ -271,7 +280,7 @@ match_containing_srcrefs <- function(l, r) {
   while (li <= nrow(ldf) && ri <= nrow(rdf)) {
     # if filenames don't match, jump to filename
     if ((t <- basename(ldf[[li, "srcfile"]])) != basename(rdf[[ri, "srcfile"]])) {
-      p <- Position(function(i) i == t, basename(rdf[ri:nrow(rdf), "srcfile"]))
+      p <- Position(function(i) identical(i, t), basename(rdf[ri:nrow(rdf), "srcfile"]))
```
Comment (maksymiuks) [inline R/srcref_df.R:283]: Wow! Really neat solution :D I like it

**Comment by maksymiuks** [comment]
> Nevertheless, please take a look, @maksymiuks and let me know if these changes resolve the issue you found.

I've checked the version from the branch against the `geepack` package that initially revealed the problem. It works perfectly fine now!

File: R/srcref_df.R (hunk 1 above)
Comment (maksymiuks) [reply R/srcref_df.R:269]: I agree it seems to be a rare case although it is good that it has been identified and we are aware that it can happen. What about keeping this in mind (like this TODO) and monitoring whether it needs addressing. Perhaps the rarity of this potential bug may cause that we will never end up in a situation when this gonna need addressing.

**Review by maksymiuks — APPROVED** [review]

File: R/srcref_df.R (hunk 1 above)
Comment (dgkf) [reply R/srcref_df.R:269]: I think we can create some heuristics to capture some parent directories in the file path comparison. Then we could differentiate

```
/home/user/R/lib/package/R/examples.R
/home/user/R/lib/dependency/inst/examples.R
```

while still treating

```
/home/user/R/lib/package/R/examples.R
/tmp/r_1234/R_LIB_covr_temp/package/R/examples.R
```

as stemming from the same code. 

As a start, taking the last 3 file path elements (`package/R/example.R`) should capture most common R package structure without introducing too much complexity.
